typed-racket-lib/typed-racket/types/prop-ops.rkt

```diff
@@ -166,30 +139,32 @@
 ;; any values of -ff are removed.
 (define/cond-contract (compact-or-props props)
   ((c:listof Prop?) . c:-> . (c:listof Prop?))
-  
-  (define-values (pos neg others)
-    (for/fold ([pos '()] [neg '()] [others '()])
-              ([prop (in-list props)])
-      (match prop
-        [(TypeProp: o t)
-         (values (union-update pos o t) neg others)]
-        [(NotTypeProp: o t)
-         (values pos (intersect-update neg o t) others)]
-        [_ (values pos neg (cons prop others))])))
 
+  (define ts+ (make-hash))
+  (define ts- (make-hash))
+  (define others '())
   
-  (let ([pos (for*/list ([p (in-mlist pos)]
-                         [p (in-value (-is-type (mcar p) (mcdr p)))]
-                         #:when (not (FalseProp? p)))
-               p)]
-        [neg (for*/list ([p (in-mlist neg)]
-                         [p (in-value (-not-type (mcar p) (mcdr p)))]
-                         #:when (not (FalseProp? p)))
-               p)])
-    (if (or (ormap TrueProp? pos)
-            (ormap TrueProp? neg))
-        (list -tt)
-        (append pos neg others))))
+  (for ([prop (in-list props)])
+    (match prop
+      [(TypeProp: o t) (hash-update! ts+ o (∪ t) -Bottom)]
+      [(NotTypeProp: o t) (hash-update! ts- o (∩ t) Univ)]
+      [_ (set! others (cons prop others))]))
+
+
+  (define pos-props
+    (for*/list ([(o t) (in-hash ts+)]
+                [p (in-value (-is-type o t))]
+                #:when (not (FalseProp? p)))
+      p))
+  (define neg-props
+    (for*/list ([(o t) (in-hash ts-)]
+                [p (in-value (-not-type o t))]
+                #:when (not (FalseProp? p)))
+      p))
+  (if (or (ormap TrueProp? pos-props)
```

Comment:
I don't think these can be reasonably combined. I did, however, swap the for/lists to be for/folds (no reverse call) and put a check at the beginning to see if the input prop list is even long enough to warrant all of this potential simplification (Or's really don't get that large in our system, we get lots of small ones). I think that's a reasonable tidying up of this function (I'll push those changes a bit).